app/modules/pwm2.c

```diff
@@ -0,0 +1,553 @@
+    {
+      setup_pin(moduleData, pin, freqFractions, freq, resolution, initDuty);
+    }
+    else
+    {
+      setup_pin(moduleData, pin, freq, freqFractions, resolution, initDuty);
+    }
+  }
+  return 0;
+}
+
+static int lpwm2_setup_pin_hz(lua_State *L)
+{
+  return lpwm2_setup_pin_common(L, true);
+}
+
+static int lpwm2_setup_pin_sec(lua_State *L)
+{
+  return lpwm2_setup_pin_common(L, false);
+}
+
+static int lpwm2_set_duty(lua_State *L)
+{
+  int pos = 0;
+  while (true)
+  {
+    LUA_INTEGER pin = luaL_optinteger(L, ++pos, -1);
+    if (pin == -1)
+    {
+      break;
+    }
+    if (pin < 1 || pin > GPIO_PIN_NUM)
+    {
+      return luaL_error(L, "pwm2 : expected 0 < pin < %d but got instead %d", GPIO_PIN_NUM, pin);
+    }
+    LUA_INTEGER duty = luaL_optinteger(L, ++pos, -1);
+    if (duty < 0 || duty > moduleData->setupData.pin[pin].pulseResolutions)
+    {
+      return luaL_error(L, "pwm2 : for pin=%d expected 0 <= initial duty <= %d but got %d", pin, moduleData->setupData.pin[pin].pulseResolutions, duty);
+    }
+    if (!isPinSetup(moduleData, pin))
+    {
+      return luaL_error(L, "pwm2 : pin=%d is not setup yet", pin);
+    }
+    set_duty(moduleData, pin, duty);
+  }
+  return 0;
+}
+
+static int lpwm2_release_pin(lua_State *L)
+{
+  if (moduleData->setupData.isStarted)
+  {
+    return luaL_error(L, "pwm2 : pwm is started, stop it first.");
+  }
+  int pos = 0;
+  while (true)
+  {
+    LUA_INTEGER pin = luaL_optinteger(L, ++pos, -1);
+    if (pin == -1)
+    {
+      break;
+    }
+    if (pin < 1 || pin > GPIO_PIN_NUM)
+    {
+      return luaL_error(L, "pwm2 : expected 0 < pin < %d but got instead %d", GPIO_PIN_NUM, pin);
+    }
+    moduleData->setupData.pin[pin].pulseResolutions = 0;
+    moduleData->interruptData.pin[pin].gpioMask = 0;
+  }
+  return 0;
+}
+
+static int lpwm2_stop(lua_State *L)
+{
+  if (!moduleData->setupData.isStarted)
+  {
+    return 0;
+  }
+  platform_hw_timer_close(PWM2_TMR_OWNER);
+  GPIO_REG_WRITE(GPIO_ENABLE_W1TC_ADDRESS, moduleData->interruptData.enabledGpioMask); // clear pins of being gpio output
+  moduleData->setupData.isStarted = false;
+  return 0;
+}
+
+static void configureAllPinsAsGpioOutput(pwm2_module_data_t *moduleData)
+{
+  for (int i = 1; i < GPIO_PIN_NUM; i++)
+  {
+    if (isPinSetup(moduleData, i))
+    {
+      PIN_FUNC_SELECT(pin_mux[i], pin_func[i]); // set pin as gpio
+      PIN_PULLUP_EN(pin_mux[i]);                // set pin pullup on
+    }
+  }
+}
+
+static void resetPinCounters(pwm2_module_data_t *moduleData)
+{
+  for (int i = 1; i < GPIO_PIN_NUM; i++)
+  {
+    if (isPinSetup(moduleData, i))
+    {
+      moduleData->interruptData.pin[i].currentInterruptCounter = 0;
+    }
+  }
+}
+
+static int lpwm2_start(lua_State *L)
+{
+  if (moduleData->setupData.isStarted)
+  {
+    return 0;
+  }
+  if (!platform_hw_timer_init(PWM2_TMR_OWNER, FRC1_SOURCE, TRUE))
+  {
+    return luaL_error(L, "pwm2: currently platform timer1 is being used by another module.\n");
```

Comment:
In fact the platform_hw_timer stuff now supports multiple users and this will only fail if it fails to allocate memory.

Reply:
right, i see. i used the older version and didn't realized this was cooked up. 
i'd have to re-test and benchmark against this version before i can comment further on the timings.
with older version i struggled to compensate interrupt handler overhead when working in non-autoload mode. that's why i opted for autoloaded mode, which seemed to interrupt rather accurate.
with the new code, i gather hw_timer_isr_cb is the actual orchestrator of interrup delays, right? how is the compensation from varying interrupt handler execution handled?

Reply:
@pjsg : i've ran same measurements against old and new platform_hw_timer code.
internal scheduling overhead for one interrupt, in average in cpu ticks, cpu80mhz: 
- previous version : for autoload mode=78, for non-autoload mode=157
- this version : for autoload mode=227, for non-autoload mode=508

internal pwm2 interrupt handler overhead is ~161+12*pins.
before total interrupt cost for 1pin was 78+161+12=251 cpu ticks, or ~318kHz max interrupt freq. 
with current code it drops to ~200kHz. 

i guess it is to be expected but frankly i'm not feeling excited. 
would you be ok if i bypass platform_hw_timer and hook directly to the interrupt (as it used to be before)?

Reply:
You can certainly do that -- can you make sure that the build fails if PWM2 is included and any of the other timer related code is included? E.g. maybe add a #define HW_TIMER_EXCLUSIVE 1 if PWM2 is included and then don't compile in the platform_hw_timer functions if HW_TIMER_EXCLUSIVE is set. [Need to think about how to make this work with the travis build which turns on all modules]. 

We had to add support for multiple users of the hw_timer as people didn't read the documentation closely enough and would complain that using module A and module B at the same time would not work. 

Does this make sense?

Reply:
absolutely. let me think a bit and propose something for review with one of next updates.

Reply:
thank you....

Reply:
could you pls take a look at the latest version and let me know if you like the approach i've taken?

Reply:
Yes -- your new approach is fine. With a bit of luck, if there is a conflict then something will fail hard!!